File: app-ios/VelaWallet/VelaWallet/Features/Signing/TrustedSigner/TrustedSignerChannel.swift
//
//  TrustedSignerChannel.swift
//  VelaWallet
//
//  The phone's end of the Trusted Signer's channel (specs 071 and 076): the
//  request goes out in the URL the page is opened with, and the answer comes
//  back as a `velawallet://sign-result?…` the system hands this app.
//
//  Everything that decides anything is the core's. This file builds no URL of
//  its own (`trustedSignerUrlLaunch`), reads no callback of its own
//  (`trustedSignerParseCallback`) and judges no answer of its own
//  (`trustedSignerVerify`, `trustedSignerVerifyCeremony`). It owns a clock, a
//  one-time token and the bookkeeping of who is waiting — and nothing else.
//
//  ## Why a scheme, and not the loopback socket it replaces
//
//  Two measured reasons, not a preference:
//
//  - the page published under spec 076 carries `default-src 'none'` INSIDE
//    the bytes that are hashed, so a page whose hash this wallet accepts
//    cannot open a WebSocket at all — measured in a real browser, the
//    listening server saw no byte. A NAVIGATION is not governed by that CSP,
//    also measured.
//  - an iOS app is suspended the moment `SFSafariViewController` covers it,
//    so it could not accept a loopback connection anyway. The system delivers
//    a scheme to a suspended app; nothing delivers a TCP connection to one.
//
//  The owner cut the loopback on 2026-09-23 (「回环 WebSocket 不做呀，现在
//  就是纯 custom schema」).
//
//  ## What the transport is not
//
//  Any app may register `velawallet://`, and which one wins is undefined on
//  iOS. So this carries an answer and authorises nothing: the core accepts
//  only a `webauthn.get` over exactly the digest this wallet computed, by a
//  credential it holds. What an interception costs is availability (the
//  wallet waits) and confidentiality (the interceptor learns the assertion).
//
//  ## One visit per request (spec 075 over this transport)
//
//  A flow — create then the member proof, a sign-in then recovery's two
//  proofs — is still one flow, ended by `end()`. But a URL carries exactly
//  one request, so each request is its own page visit with its own one-time
//  token, and `send(_:)` opens the page again rather than writing down a
//  socket that is not there.
//

import Foundation
import Network
import VelaCore

/// One request of a session: a passkey ceremony (spec 075) or a signature
/// (071). Both ride the same channels; only who judges the answer differs,
/// and that judge is the core's.
enum TrustedSignerAsk {
    /// `request` is `trustedSignerCeremonyRequest`'s, `operationJson` the
    /// machine operation's own wire JSON, `memberChallenge` the registry
    /// challenge the WALLET fetched (a member proof only).
    case ceremony(request: String, operationJson: String, memberChallenge: Data?)
    /// `request` is `trustedSignerRequest`'s; the answer must sign `digest`
    /// with one of `keys`.
    case signature(request: String, digest: Data, keys: [WalletKeyRecord])
}

/// The page this flow is talking to. One implementation — the cross-device
/// ones were retired by the owner on 2026-09-23 — kept as a protocol because
/// it is what the driver holds while a flow is open.
protocol TrustedSignerConversation: AnyObject {
    /// The flow's next request: a new visit to the page, with a new one-time
    /// token.
    func send(_ ask: TrustedSignerAsk) async -> TrustedSignerChannel.Ending
    /// The flow is over.
    func end()
}

/// The page's refusal codes in the core's vocabulary.
///
/// What used to live here — building the `intent` envelope, reading a
/// `result` frame — belonged to the socket channels and went with them. On
/// the custom scheme the core writes the whole launch URL and reads the whole
/// callback, so a shell has nothing left to shape.
enum TrustedSignerAnswer {

    /// The page's `code`, in the core's vocabulary — `trusted_signer::refusal`,
    /// which the bindings do not export on its own. A signing answer is the
    /// only place a shell needs it: a ceremony's `error` goes to
    /// `trustedSignerVerifyCeremony`, which reads the code itself.
    static func refusal(code: String) -> TrustedSignerRefusal {
        code.isEmpty || code == "user_rejected" ? .declined : .pageRefused(code: code)
    }
}

final class TrustedSignerChannel: TrustedSignerConversation {

    /// How one REQUEST of a flow ended.
    enum Ending: Equatable {
        /// The core's verdict on a signature: an answer it verified, or why
        /// there is none — `declined` for a page that went away holding the
        /// request, and for a cancel.
        case outcome(TrustedSignerOutcome)
        /// Spec 075: the core's verdict on a passkey ceremony.
        case ceremony(TrustedSignerCeremonyOutcome)
        /// Nobody answered in time. The core has no clock; this is the shell's.
        case timedOut
        /// The request could not be put to a page at all.
        case unavailable
    }

    /// The contract's five minutes — idle, so a flow of several requests is
    /// not cut off halfway (contract §1.5).
    static let defaultTimeout: TimeInterval = 5 * 60

    let signerUrl: String
    /// The CURRENT visit's one-time token. A new one per request, so
    /// "one-time" is literally true and no answer can be read as another
    /// request's.
    private(set) var token: String
    /// The URL the current visit was opened with — what "open the page again"
    /// re-opens, unchanged, token and all.
    private(set) var launchUrl: URL?

    /// The request in flight. A refusal is shaped like the request that
    /// earned it, so a ceremony's decline never arrives as a signing verdict
    /// (and the spine never reads one as a shell bug).
    private var current: TrustedSignerAsk
    private let timeout: TimeInterval

    /// Opens a page. Set by whoever owns the tab, because presenting is its
    /// business and not this class's.
    var openPage: ((URL) -> Void)?

    /// Terminal: a timeout, a cancel, or a request that could not go out.
    /// Answered to every later caller.
    private var terminal: Ending?
    /// The flow has been shut down (`end()`), with no terminal verdict.
    private var closed = false
    /// A verdict nobody was waiting for yet.
    private var pendingEndings: [Ending] = []
    private var waiting: [CheckedContinuation<Ending, Never>] = []
    private var deadline: DispatchWorkItem?

    init(
        signerUrl: String,
        first: TrustedSignerAsk,
        timeout: TimeInterval = TrustedSignerChannel.defaultTimeout
    ) {
        self.signerUrl = signerUrl
        self.current = first
        self.timeout = timeout
        self.token = Self.freshToken()
    }

    /// The 071 door: one signature, one page visit.
    convenience init(
        signerUrl: String,
        requestJson: String,
        digest: Data,
        keys: [WalletKeyRecord],
        timeout: TimeInterval = TrustedSignerChannel.defaultTimeout
    ) {
        self.init(
            signerUrl: signerUrl,
            first: .signature(request: requestJson, digest: digest, keys: keys),
            timeout: timeout
        )
    }

    /// Whether anything more can happen here.
    private var stopped: Bool { closed || terminal != nil }

    // MARK: - Putting a request

    /// The first request: build its URL, start waiting for its callback, and
    /// hand the URL back so the caller can open the page. `nil` when the core
    /// would not build one — a request that is not JSON, which is a bug in
    /// this app rather than anything a person did.
    func start() -> URL? {
        guard !stopped, let url = visit(current) else {
            finish(.unavailable)
            return nil
        }
        armDeadline()
        return url
    }

    /// The flow's next request. On this transport that is a NEW visit: a new
    /// token, a new URL, and the page opened again — there is no socket to
    /// write down.
    func send(_ ask: TrustedSignerAsk) async -> Ending {
        guard !stopped else { return .unavailable }
        TrustedSignerCallbacks.forget(token)
        token = Self.freshToken()
        current = ask
        guard let url = visit(ask) else { return .unavailable }
        armDeadline()
        openPage?(url)
        return await ending()
    }

    /// Build this visit's URL and register for its callback.
    private func visit(_ ask: TrustedSignerAsk) -> URL? {
        let requestJson = switch ask {
        case .ceremony(let request, _, _): request
        case .signature(let request, _, _): request
        }
        guard let launch = try? trustedSignerUrlLaunch(
            base: signerUrl, requestJson: requestJson, token: token
        ), let url = URL(string: launch) else { return nil }
        launchUrl = url
        TrustedSignerCallbacks.await(token) { [weak self] callback in
            self?.answered(callback)
        }
        return url
    }

    /// Waits for the CURRENT request's verdict. A cancelled task cancels the
    /// ceremony — the send screen's cancel is a `Task` cancel, and it must
    /// not leave a request waiting on a callback that will never come.
    func ending() async -> Ending {
        if let next = take() { return next }
        return await withTaskCancellationHandler {
            await withCheckedContinuation { continuation in
                if let next = take() {
                    continuation.resume(returning: next)
                } else {
                    waiting.append(continuation)
                }
            }
        } onCancel: {
            Task { @MainActor [weak self] in self?.cancel() }
        }
    }

    /// The verdict already in hand, if there is one.
    private func take() -> Ending? {
        if !pendingEndings.isEmpty { return pendingEndings.removeFirst() }
        if let terminal { return terminal }
        return closed ? .unavailable : nil
    }

    // MARK: - The answer

    /// A `velawallet://sign-result?…` for THIS visit's token. The core reads
    /// it and the core judges what it carried.
    private func answered(_ callback: URL) {
        guard !stopped else { return }
        let read = trustedSignerParseCallback(url: callback.absoluteString, token: token)
        switch read {
        case .answered(let answerJson):
            // Unwrapped, because that is what the URL channel carries: a
            // signature's `result` object, or a ceremony's own envelope
            // (`intake.js`: `body.result !== undefined ? body.result : body`).
            switch current {
            case .signature(_, let digest, let keys):
                deliver(.outcome(trustedSignerVerify(
                    resultJson: answerJson, digest: digest, keys: keys
                )))
            case .ceremony(_, let operationJson, let memberChallenge):
                deliver(.ceremony(trustedSignerVerifyCeremony(
                    operationJson: operationJson, answerJson: answerJson,
                    signerOrigin: signerUrl, expectedMemberChallenge: memberChallenge
                )))
            }
        case .refused(let refusal):
            deliver(refused(refusal))
        }
    }

    // MARK: - Endings

    /// A refusal shaped like the request that earned it.
    static func declined(for ask: TrustedSignerAsk) -> Ending {
        refused(.declined, for: ask)
    }

    static func refused(_ refusal: TrustedSignerRefusal, for ask: TrustedSignerAsk) -> Ending {
        switch ask {
        case .ceremony: .ceremony(.refused(refusal: refusal))
        case .signature: .outcome(.refused(refusal: refusal))
        }
    }

    /// The same, for the request in flight.
    private func refused(_ refusal: TrustedSignerRefusal) -> Ending {
        Self.refused(refusal, for: current)
    }

    /// The flow is done.
    func end() {
        guard !stopped else { return }
        closed = true
        shutDown()
        let waiters = waiting
        waiting.removeAll()
        waiters.forEach { $0.resume(returning: .unavailable) }
    }

    /// The waiting sheet's cancel: the person declined (contract §2).
    func cancel() {
        finish(refused(.declined))
    }

    /// The person closed the tab.
    ///
    /// On this transport that is NOT an answer. A page can be closed after it
    /// has navigated to the callback — the tab often outlives the navigation
    /// — so reading a dismissal as "closed without signing" would turn a
    /// signature the person just approved into a refusal. The callback that
    /// is on its way settles the request; if none comes, the clock does, and
    /// the waiting sheet keeps its two ways on in the meantime.
    func pageClosed() {}

    /// One request's verdict, to whoever is waiting for it — or kept for the
    /// caller that has not asked yet.
    private func deliver(_ ending: Ending) {
        guard !stopped else { return }
        armDeadline()
        if waiting.isEmpty {
            pendingEndings.append(ending)
        } else {
            waiting.removeFirst().resume(returning: ending)
        }
    }

    /// A terminal ending: the first one wins, and everything closes with it.
    private func finish(_ ending: Ending) {
        guard !stopped else { return }
        terminal = ending
        shutDown()
        let waiters = waiting
        waiting.removeAll()
        waiters.forEach { $0.resume(returning: ending) }
    }

    private func shutDown() {
        deadline?.cancel()
        deadline = nil
        TrustedSignerCallbacks.forget(token)
    }

    // MARK: - The clock

    /// The idle clock, restarted. A flow of several requests is cut off only
    /// when nothing has moved for the whole window (contract §1.5).
    private func armDeadline() {
        deadline?.cancel()
        guard !stopped else { return }
        let deadline = DispatchWorkItem { [weak self] in
            MainActor.assumeIsolated { self?.finish(.timedOut) }
        }
        self.deadline = deadline
        DispatchQueue.main.asyncAfter(deadline: .now() + timeout, execute: deadline)
    }

    // MARK: - The token

    /// 16 bytes from the system's CSPRNG, base64url without padding.
    static func freshToken() -> String {
        var generator = SystemRandomNumberGenerator()
        let bytes = (0..<16).map { _ in UInt8.random(in: .min ... .max, using: &generator) }
        return Data(bytes).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

/// The callbacks this app is waiting for, by the one-time token they name.
///
/// App-wide rather than held by a view, because the two ends are in different
/// places: a request waits inside whatever task asked for a signature, and the
/// callback arrives at `onOpenURL` on the root view.
///
/// Keyed by the token, and only the token: it is what says WHICH request a
/// callback belongs to. A callback for a token nobody awaits reaches nothing
/// and is dropped in silence — no route change, no state change, no error a
/// person cannot act on.
@MainActor
enum TrustedSignerCallbacks {

    private static var waiting: [String: (URL) -> Void] = [:]

    static func await(_ token: String, _ sink: @escaping (URL) -> Void) {
        waiting[token] = sink
    }

    static func forget(_ token: String) {
        waiting.removeValue(forKey: token)
    }

    /// A `velawallet://…` the system handed this app.
    ///
    /// **It is an event for a pending request, not a navigation.** `true` when
    /// it settled one, and `false` for every other URL — the wallet's own
    /// `pay` and `open` links included — which the caller then routes as it
    /// always did. Nothing here touches the screen either way.
    @discardableResult
    static func deliver(_ url: URL) -> Bool {
        guard let token = trustedSignerCallbackToken(url: url.absoluteString),
              let sink = waiting.removeValue(forKey: token)
        else { return false }
        sink(url)
        return true
    }
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/signing/trustedsigner/TrustedSignerChannel.kt
package app.getvela.wallet.feature.signing.trustedsigner

import app.getvela.wallet.core.diagnostics.VelaLog
import app.getvela.wallet.feature.onboarding.core.Assertion
import app.getvela.wallet.feature.onboarding.core.FailureKind
import app.getvela.wallet.feature.onboarding.core.PasskeyFailure
import app.getvela.wallet.feature.send.core.TrustedSigner
import app.getvela.wallet.feature.send.core.TrustedSignerLabels
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull
import uniffi.vela_core_uniffi.TrustedSignerCeremonyOutcome
import uniffi.vela_core_uniffi.TrustedSignerOutcome
import uniffi.vela_core_uniffi.TrustedSignerRefusal
import uniffi.vela_core_uniffi.WalletKeyRecord
import java.security.SecureRandom

/**
 * The Trusted Signer, as the rest of the app sees it (specs 071 and 075).
 *
 * It is the fourth passkey route: a page the person reads, which checks the
 * request and runs the WebAuthn ceremony itself. Everything that can ask for
 * one — a send, a dApp request, the key backup, and since 075 creating a
 * wallet, signing in and every proof inside those flows — comes through here.
 *
 * The page is a Custom Tab over the app, answering through the custom scheme
 * ([TrustedSignerScheme]). That is the only channel: the owner retired the
 * cross-device ones on 2026-09-23 ("客户端支持回环 + 蓝牙就够了" and then
 * "我确定砍掉蓝牙"), because only a page THIS device fetched can be checked
 * against what it is supposed to be — and then the loopback socket with them
 * (「回环 WebSocket 不做呀，现在就是纯 custom schema」), because the published
 * page's own `default-src 'none'` makes a socket from inside it impossible.
 *
 * Two decisions live in this class and nowhere else:
 *
 * 1. **Who owns the visit.** A create mints a key and then confirms its
 *    membership; a recovery signs twice. Those are one FLOW, and [endFlow]
 *    ends it — though on the custom-scheme channel each request is its own
 *    page visit, because a URL carries exactly one request.
 * 2. **What a refusal means to the caller.** For a SIGNATURE, nothing signed
 *    is a cancelled ceremony whatever the reason (071 contract §5), so the
 *    request stays open and can be signed another way. For a CEREMONY inside
 *    create or sign-in, a closed page is a cancellation and everything else
 *    carries its own sentence, which is what the machines' failure shapes
 *    already distinguish.
 */
class TrustedSignerChannel(
    /** The page the wallet opens by default (`sign_pref`'s, always usable). */
    private val signerUrl: () -> String,
    /** Open [url] in a Custom Tab over the app; `false` when nothing is on screen to open it from. */
    private val openPage: (url: String) -> Boolean,
    /** Put the app back over the tab. */
    private val bringBack: () -> Unit,
    private val words: () -> Words,
    /** The chain's name and coin, the account's name — from the wallet's own lists. */
    private val labels: (chainId: Int, account: String) -> TrustedSignerLabels = { _, _ -> TrustedSignerLabels() },
    private val timeoutMs: Long = 5 * 60_000L,
    private val random: SecureRandom = SecureRandom(),
    /** This app's own mark, inline PNG, for the page to draw beside the name. */
    private val appIcon: () -> String = { "" },
    /** `Vela Wallet <version>` — what the page shows as the requester, marked there as self-reported. */
    private val appName: String = "vela-android",
) : TrustedSigner {

    /** The sentences a refusal is told in (`componentsUi.signing.trustedSigner*`). */
    data class Words(
        val closed: String,
        val refused: String,
        val mismatch: String,
        val timeout: String,
    )

    sealed interface State {
        data object Idle : State

        /** The page is open (or being opened) at [url] and the socket is listening. */
        data class Waiting(val url: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state

    /** Why the last attempt did not sign — shown until the next one starts. */
    val notice = MutableStateFlow<String?>(null)

    private val one = Mutex()

    /** The flow's conversation with one page; `null` between flows. */
    @Volatile
    private var wire: TrustedSignerWire? = null

    /**
     * Why this flow's channel could not be opened at all, when there is a
     * better sentence than "the page was closed".
     *
     * A request that never opened a page comes back as [TrustedSignerAnswer.Cancelled]
     * like any other, and the caller is told "the Trusted Signer was closed
     * without signing" — which is true of a dismissed sheet and a lie about a
     * radio the person was never allowed to turn on. This carries the real
     * reason as far as the sentence, and is cleared at the start of every
     * request so it can never outlive the attempt that set it.
     */
    @Volatile
    private var unopenable: String? = null

    override fun describe(chainId: Int, account: String): TrustedSignerLabels = labels(chainId, account)

    // -- what the screens drive -----------------------------------------------

    /** Cancel, on the waiting sheet: the same as closing the page. */
    fun cancel() {
        wire?.cancel()
    }

    /** "Open the page again" — same port, same token. */
    fun reopen() {
        wire?.reopen()
    }

    /**
     * The flow is over — a create finished or failed, a sign-in settled, the
     * person left the screen. Says `bye`, lets the page go and brings the app
     * back. Safe to call when no flow is open.
     */
    fun endFlow() {
        val live = wire
        wire = null
        cancel()
        live?.end()
        _state.value = State.Idle
        bringBack()
    }

    // -- the two things a caller can ask for ----------------------------------

    /**
     * Spec 071: sign [digest] with one of [keys]. [signerOrigin] is the page
     * the key lives behind, empty for the person's own page.
     *
     * A standalone signature is a flow of one: the page opens, signs and goes.
     * Inside a create or a sign-in the conversation is already open, and this
     * simply puts the next request down it.
     */
    override suspend fun sign(
        requestJson: String,
        digest: ByteArray,
        keys: List<WalletKeyRecord>,
        signerOrigin: String,
    ): Assertion {
        // A signature asked for on its own opens and closes its own flow; one
        // asked for inside a create or a sign-in belongs to that flow's visit.
        // WHICH it is can only be decided under the lock — two callers reading
        // `wire` from outside it would both think they owned the flow, or
        // neither would.
        val put = put(TrustedSignerAsk.Signature(requestJson, digest, keys), signerOrigin)
        try {
            val answer = put.answer
            val outcome = (answer as? TrustedSignerAnswer.Signed)?.outcome
                // Nothing signed: a cancelled ceremony, so the request stays open.
                ?: refuse(FailureKind.Cancelled, sentenceFor(answer))
            return when (outcome) {
                is TrustedSignerOutcome.Accepted -> Assertion(
                    credentialIdHex = outcome.credentialIdHex,
                    signatureDerHex = hex(outcome.assertion.signatureDer),
                    authenticatorDataHex = hex(outcome.assertion.authenticatorData),
                    clientDataJsonHex = hex(outcome.assertion.clientDataJson),
                    userIdHex = null,
                    authenticatorAttachment = "",
                )
                is TrustedSignerOutcome.Refused -> refuse(FailureKind.Cancelled, told(outcome.refusal))
            }
        } finally {
            if (put.ownsFlow) endFlow()
        }
    }

    /**
     * Spec 075: run a passkey ceremony on the page. [operationJson] is the
     * machine operation's own wire JSON; the answer comes back already judged
     * by the core, ready to be reported as the platform ceremony's result.
     *
     * The conversation STAYS OPEN: a create's member proof and a recovery's
     * second signature are the next request of the same visit. [endFlow] ends
     * it.
     *
     * A refusal throws a [PasskeyFailure] — the vocabulary the executor already
     * answers every ceremony's failure in.
     */
    suspend fun ceremony(
        requestJson: String,
        operationJson: String,
        expectedMemberChallenge: ByteArray? = null,
        signerOrigin: String = "",
    ): TrustedSignerCeremonyOutcome {
        val answer = put(
            TrustedSignerAsk.Ceremony(requestJson, operationJson, expectedMemberChallenge),
            signerOrigin,
        ).answer
        val outcome = (answer as? TrustedSignerAnswer.Ceremonial)?.outcome
            ?: refuse(kindOf(answer), sentenceFor(answer))
        if (outcome is TrustedSignerCeremonyOutcome.Refused) {
            refuse(
                if (outcome.refusal == TrustedSignerRefusal.Declined) {
                    FailureKind.Cancelled
                } else {
                    FailureKind.Other
                },
                told(outcome.refusal),
            )
        }
        return outcome
    }

    // -- the machinery --------------------------------------------------------

    /** One request's verdict, and whether this caller opened the flow it rode. */
    private class Put(val answer: TrustedSignerAnswer, val ownsFlow: Boolean)

    private suspend fun put(ask: TrustedSignerAsk, signerOrigin: String): Put =
        one.withLock {
            notice.value = null
            unopenable = null
            val existing = wire
            val mine = existing == null
            val live = existing ?: open(signerOrigin) ?: return@withLock Put(TrustedSignerAnswer.Cancelled, true)
            val answer = try {
                live.ask(ask)
            } catch (cancellation: CancellationException) {
                // The caller went away mid-request: a screen left, a ViewModel
                // cleared, a scope torn down. Nothing was signed, and a page
                // left open would hold a bound socket and a tab in front of an
                // app that is no longer asking for anything.
                endFlow()
                throw cancellation
            }
            if (answer !is TrustedSignerAnswer.Signed && answer !is TrustedSignerAnswer.Ceremonial) {
                // A flow that did not answer has no session left to reuse.
                endFlow()
            } else {
                // Answered. On the custom-scheme channel a visit ENDS with its
                // answer — there is no session to keep, and the next request
                // opens the page again — so the person must be brought back to
                // the wallet now.
                //
                // Leaving the tab up is what the socket channel did, because
                // there the next request arrived on the same connection and the
                // page itself moved on. Here nothing moves the page, and the
                // owner found exactly that: 「签完名后，会卡在 完成，结果已交回
                // 钱包」 — the page says it handed the answer back and the
                // wallet, which had it, was behind the tab.
                _state.value = State.Idle
                bringBack()
            }
            Put(answer, mine)
        }

    /** Open the flow's conversation: a Custom Tab, answering by custom scheme. */
    private fun open(signerOrigin: String): TrustedSignerWire? {
        val page = signerOrigin.ifEmpty { signerUrl() }
        // A channel that cannot even be built is an unopened page, not an
        // exception on its way through the signing paths.
        val built = runCatching {
            TrustedSignerScheme(
                base = page,
                openPage = openPage,
                timeoutMs = timeoutMs,
                random = random,
                onOpened = { url -> _state.value = State.Waiting(url) },
            )
        }.getOrElse { error ->
            VelaLog.failure("trustedsigner", "the channel could not be opened", error)
            _state.value = State.Idle
            return null
        }
        wire = built
        return built
    }

    private fun kindOf(answer: TrustedSignerAnswer): FailureKind = when (answer) {
        TrustedSignerAnswer.Cancelled -> FailureKind.Cancelled
        else -> FailureKind.Other
    }

    /** Why a request came back with nothing, in the person's language. */
    private fun sentenceFor(answer: TrustedSignerAnswer): String {
        val w = words()
        return when (answer) {
            TrustedSignerAnswer.TimedOut -> w.timeout
            is TrustedSignerAnswer.Unreachable -> {
                VelaLog.event(
                    "trustedsigner",
                    "the channel could not carry the request",
                    "why" to answer.why.name,
                    "detail" to answer.detail,
                )
                w.timeout
            }
            is TrustedSignerAnswer.Ceremonial -> told(
                (answer.outcome as? TrustedSignerCeremonyOutcome.Refused)?.refusal
                    ?: TrustedSignerRefusal.Declined,
            )
            is TrustedSignerAnswer.Signed -> told(
                (answer.outcome as? TrustedSignerOutcome.Refused)?.refusal
                    ?: TrustedSignerRefusal.Declined,
            )
            // A channel that never opened says why; everything else that
            // comes back with nothing is a page the person closed.
            TrustedSignerAnswer.Cancelled -> unopenable ?: w.closed
        }
    }

    private fun told(refusal: TrustedSignerRefusal): String {
        val w = words()
        return when (refusal) {
            TrustedSignerRefusal.Declined -> w.closed
            is TrustedSignerRefusal.PageRefused -> w.refused
            else -> {
                // Signed, but not what this wallet asked for — logged, never used.
                VelaLog.event("trustedsigner", "answer refused", "refusal" to refusal.toString())
                w.mismatch
            }
        }
    }

    /**
     * Nothing was signed. The sentence rides on [notice] for the signing
     * sheet, and the failure itself is the vocabulary every caller already
     * classifies on.
     */
    private fun refuse(kind: FailureKind, sentence: String): Nothing {
        notice.value = sentence
        throw PasskeyFailure(kind, sentence)
    }

    private fun hex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }
}
